orttraining/orttraining/core/framework/distributed_run_context.cc

```diff
@@ -47,14 +45,46 @@ DistributedRunContext::DistributedRunContext(int32_t world_rank,
   params_.data_parallel_size = data_parallel_size;
   params_.horizontal_parallel_size = horizontal_parallel_size;
   params_.pipeline_stage_size = pipeline_stage_size;
-  groups_.resize(2);
+  groups_.resize(static_cast<size_t>(WorkerGroupType::WorkerGroupTypeCount));
 
+  // Consider distributed training forms three axis, data parallel, horizontal parallel and pipeline parallel.
+  // The three axis are pependicular to each other and like x, y, z axis in 3D space (but only the positive directions).
+  // Now the world ranks are numbers filling in this 3D space alone the three axis. It will try fill alone the
+  // horizontal axis first, then data parallel axis, and last pipeline parallel axis.
+  //
+  // Ranks that are aligned with a specific axis forms a group. For example, for a 3D cube with size
+  // horizontal_parallel x data_parallel x pipeline_parallel equal to 4x3x2, there are 24 rank numbers fill into the
+  // cubic. It will have 6 horizontal groups, 8 data parallel groups and 12 pipeline groups, as shown below.
+  //
+  //          pipeline
+  //            ^
+  //            | 12, 16, 20,
+  //            |  13, 17, 21,
+  //            |   14, 18, 22,
+  //            |    15, 19, 23,
+  //            |__________________> data
+  //             \ 0, 4, 8,
+  //              \ 1, 5, 9,
+  //               \ 2, 6, 10,
+  //                \ 3, 7, 11,
+  //                 v
+  //                horizontal
+  //
+  // For a given world rank, say 11, it will be in the 3th data parallel group (3, 7, 11), with in-group index 2,
```

Comment:
>(3, 7, 11) [](start = 79, length = 10)

Not sure if it's (8, 9, 10, 11)?

Reply:
numbers aligned with horizontal axis is the horizontal group. So (8, 9, 10, 11) form a horizontal group. (It matches the horizontal parallel size 4 too.)

Reply:
Are GPUs with different horizontal coordinate in the same horizontal group? In this coordinate system, (3, 7, 11) have the same horizontal coordinate but different data coordinates, so I feel (3, 7, 11) is horizontal group.

---
In reply to: [462639398](https://github.com/microsoft/onnxruntime/pull/4648#discussion_r462639398) [](ancestors = 462639398)

Reply:
Nah, as mentioned in the comment, ranks that are aligned with a certain axis forms a group in that particular parallelism. It is not related to coordinates. (Also, coordinate is not a good way to describe groups in this case, as (15, 19, 23) also has the same horizontal coordinate as (3, 7, 11) in this example.) So the best way to depict groups is viewing them as axis-aligned groups.